### PR TITLE
[Backport] Throw when reader is completed with an exception

### PR DESCRIPTION
**Pull request by BrennanConroy**
Backport of https://github.com/dotnet/runtime/pull/43776

## Customer Impact

While fixing a usability concern with pipes we introduced this regression. Customers expecting an exception from a completed reader will instead get a completion without an exception.

## Testing

We were missing a test case that was found by a customers' test suite. Test has been added and bug fixed with this change.

## Regression?

Yes, previous patch we introduced this regression during another fix.

## Risk

Low, fix is very small and easy to understand

### CODE REVIEW

File: src/System.IO.Pipelines/tests/PipeReaderWriterFacts.cs [hunk 1]
```diff
@@ -200,36 +200,53 @@ public async Task HelloWorldAcrossTwoBlocks()
             Assert.Equal(" World", Encoding.ASCII.GetString(worldBytes));
         }
 
-        [Fact]
-        public async Task ReadAsync_ThrowsIfWriterCompletedWithException()
+        [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (halter73) [inline src/System.IO.Pipelines/tests/PipeReaderWriterFacts.cs:204]: It looks like you need a using.

**Review by halter73 — APPROVED** [review]

**Comment by BrennanConroy** [comment]
Fix verified by reporter.

**Review by Anipik — APPROVED** [review]
@aik-jahoda  can you resolve the merge conflicts and get this in

**Comment by BrennanConroy** [comment]
Ping, I fixed the merge conflict. 3.1 changes are supposed to be in by today I believe. Anything else needed to merge?

**Comment by wtgodbe** [comment]
Looks good to me, merging

**Pull request merged**